glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt

```diff
@@ -112,14 +121,20 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
                         action.request.body,
                         action.request.headers,
                         Glean.configuration
-                    )
+                    ).toFfi()
+
+                    if (result == Constants.UPLOAD_RESULT_RECOVERABLE ||
+                        result == Constants.UPLOAD_RESULT_UNRECOVERABLE) {
+                            uploadFailures++
+                        }
 
                     // Process the upload response
-                    LibGleanFFI.INSTANCE.glean_process_ping_upload_response(incomingTask, result.toFfi())
+                    LibGleanFFI.INSTANCE.glean_process_ping_upload_response(incomingTask, result)
                 }
                 PingUploadTask.Wait -> return Result.retry()
                 PingUploadTask.Done -> return Result.success()
             }
-        } while (true)
+        } while (uploadFailures < MAX_RETRIES)
+        return Result.failure()
```

Comment:
Unsure if this should really be a `failure` or a `retry`.

Reply:
We use a `OneTimeWorkRequest` which defaults to exponential backoff on `retry`.
I think `failure` is the right thing here. If we fail too often, we don't want to reschedule it at all. When a ping gets submitted again (e.g. next time the app is backgrounded) any pending pings are picked up again anyway.

Reply:
cc @Dexterp37 just for a quick look

Reply:
Works for me, assuming failure doesn't mean deleting the pings

Reply:
Nope, it's only marking the worker as failed. The pings are handled inside glean-core and will be retried by the next worker.